DifferentialLineAdjustmentMap.php

```diff
@@ -0,0 +1,376 @@
+          $cursor = $line - 1;
+          while (isset($nmap[$cursor])) {
+            $prev = $nmap[$cursor];
+            $prev = reset($prev);
+            if ($prev == $to_line) {
+              $offset++;
+            } else {
+              break;
+            }
+            $cursor--;
+          }
+        }
+
+        $to_line = -$to_line;
+        if (!$this->isInverse) {
+          $deleted = true;
+        }
+      }
+      $line = $to_line;
+    } else {
+      $line = $line + $this->finalOffset;
+    }
+
+    if ($this->nextMapInChain) {
+      $chain = $this->nextMapInChain->mapLine($line, $is_end);
+      list($chain_deleted, $chain_offset, $line) = $chain;
+      $deleted = ($deleted || $chain_deleted);
+      if ($chain_offset !== false) {
+        if ($offset === false) {
+          $offset = 0;
+        }
+        $offset += $chain_offset;
+      }
+    }
+
+    return array($deleted, $offset, $line);
+  }
+
+
+  /**
+   * Build a derived map which maps deleted lines to the nearest valid line.
+   *
+   * This computes a "nearest line" map and a final-line offset. These
+   * derived maps allow us to map deleted code to the previous (or next) line
+   * which actually exists.
+   */
+  private function buildNearestMap() {
+    $map = $this->map;
+    $nmap = array();
+
+    $nearest = 0;
+    foreach ($map as $key => $value) {
+      if ($value) {
+        $nmap[$key] = $value;
+        $nearest = end($value);
+      } else {
+        $nmap[$key][0] = -$nearest;
+      }
+    }
+
+    if (isset($key)) {
+      $this->finalOffset = ($nearest - $key);
+    } else {
+      $this->finalOffset = 0;
+    }
+
+    foreach (array_reverse($map, true) as $key => $value) {
+      if ($value) {
+        $nearest = reset($value);
+      } else {
+        $nmap[$key][1] = -$nearest;
+      }
+    }
+
+    $this->nearestMap = $nmap;
+
+    return $this;
+  }
+
+  public static function newFromHunks(array $hunks) {
+    assert_instances_of($hunks, 'DifferentialHunk');
+
+    $map = array();
+    $o = 0;
+    $n = 0;
+
+    $hunks = msort($hunks, 'getOldOffset');
+    foreach ($hunks as $hunk) {
+
+      // If the hunks are disjoint, add the implied missing lines where
+      // nothing changed.
+      $min = ($hunk->getOldOffset() - 1);
+      while ($o < $min) {
+        $o++;
+        $n++;
+        $map[$o][] = $n;
+      }
+
+      $lines = $hunk->getStructuredLines();
+      foreach ($lines as $line) {
+        switch ($line['type']) {
+          case '-':
+            $o++;
+            $map[$o] = array();
+            break;
+          case '+':
+            $n++;
+            $map[$o][] = $n;
+            break;
+          case ' ':
+            $o++;
+            $n++;
+            $map[$o][] = $n;
+            break;
+          default:
+            break;
+        }
+      }
+    }
+
+    $map = self::reduceMapRanges($map);
+
+    return self::newFromMap($map);
+  }
+
+  public static function newFromMap(array $map) {
```

Comment:
Prefer the modern spelling, "mappe".

Also, this MUST not be `static` or ALL production data will be deleted FOREVER.